src/sql/workbench/services/connection/browser/connectionManagementService.ts

```diff
@@ -972,6 +1012,50 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 		return true;
 	}
 
+	/**
+	 * Refresh Azure access token
+	 * @param profile connection profile
+	 */
+	public async timedRefreshAzureAccountTokens(profile: ConnectionProfile): Promise<void> {
+
+		let uri = this.getConnectionUri(profile);
+		//wait for the pending reconnction promise if any
+		const previousReconnectPromise = this._uriToReconnectPromiseMap[uri];
+		if (previousReconnectPromise) {
+			this._logService.info(`Found pending reconnect promise for uri ${uri}, waiting.`);
+			try {
+				const previousConnectionResult = await previousReconnectPromise;
+				if (previousConnectionResult && previousConnectionResult.connected) {
+					this._logService.info(`Previous pending reconnection for uri ${uri} succeeded.`);
+					return;
+				}
+				this._logService.info(`Previous pending reconnection for uri ${uri} failed.`);
+			} catch (err) {
+				this._logService.info(`Previous pending reconnect promise for uri ${uri} is rejected with error ${err}, will attempt to reconnect if necessary.`);
+			}
+		}
+
+
+		this._logService.info(`Token refresh for connection ${profile.id} with uri ${uri}`);
+		try {
+			const connectionResultPromise = this.connect(profile, uri);
```

Comment:
Does this mean we're going to be disconnecting and re-connecting every 10min now for all Azure connections? What if the user is in the middle of executing something?

We may want to look into using a [SqlAuthenticationProvider](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlauthenticationprovider?view=netframework-4.8) to handle the refresh - my understanding is that if you create a custom one you can hook it up to have that called whenever a token is needed. So we would get a new token and refresh it and pass that into STS but then not do anything with the connection itself until a new token was requested. (@shueybubbles may be able to help here)

Reply:
1 - Stop setting `SqlConnection.AccessToken`  directly, and don't implement IRenewableToken for SMO anymore.
2 - Register a new `SqlAuthenticationProvider` implementation for the ActiveDirectoryInteractive authentication type. This implementation should return the latest token you have for the given parameters. https://docs.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlauthenticationprovider.setprovider?view=sqlclient-dotnet-standard-4.1
3 - when a user wants to use interactive AAD auth, add `ActiveDirectoryInteractive` as the `AuthenticationType` to the connection string. SqlClient will call your SqlAuthenticationProvider implementation as needed to get a token.